File: src/main/java/xf/xflp/opt/construction/onetype/OneContainerOneTypeAddPacker.java
package xf.xflp.opt.construction.onetype;

import xf.xflp.base.XFLPModel;
import xf.xflp.base.container.Container;
import xf.xflp.base.item.Item;
import xf.xflp.exception.XFLPException;
import xf.xflp.opt.Packer;

import java.util.Arrays;
import java.util.List;

/**
 * Copyright (c) 2012-2026 Holger Schneider
 * All rights reserved.
 *
 * This source code is licensed under the MIT License (MIT) found in the
 * LICENSE file in the root directory of this source tree.
 *
 * This packer puts the items in a sequence into one container with one container type.
 * Items will only be added to a container.
 *
 * @author hschneid
 *
 */
public class OneContainerOneTypeAddPacker implements Packer {

	private final Heuristic heuristic;

	public OneContainerOneTypeAddPacker(Heuristic heuristic) {
		this.heuristic = heuristic;
	}

	@Override
	public void execute(XFLPModel model) throws XFLPException {
		Container container = model.getContainerTypes()[0].newInstance();

		List<Item> unplannedItemList = heuristic
				.createLoadingPlan(
						Arrays.asList(model.getItems()),
						container
				);

		// Put result into model
		model.setContainers(new Container[]{container});
		model.setUnplannedItems(unplannedItemList.toArray(new Item[0]));
	}

}

File: src/xflp/opt/construction/onetype/one_container_one_type_add_packer.py
# Copyright (c) 2012-2025 Holger Schneider
# All rights reserved.
#
# This source code is licensed under the MIT License (MIT) found in the
# LICENSE file in the root directory of this source tree.

"""
This packer puts the items in a sequence into one container with one container type.
Items will only be added to a container.

Author: hschneid
"""

from typing import TYPE_CHECKING

from xflp.opt.packer import Packer
from xflp.opt.construction.onetype.single_bin_add_heuristic import SingleBinAddHeuristic

if TYPE_CHECKING:
    from xflp.base.xflp_model import XFLPModel


class OneContainerOneTypeAddPacker(Packer):
    """Packer for single container with one type (add only)."""

    def execute(self, model: 'XFLPModel') -> None:
        """
        Execute the packer.

        Args:
            model: The XFLP model

        Raises:
            XFLPException: If packing fails
        """
        container = model.get_container_types()[0].new_instance()

        strategy = model.get_parameter().get_preferred_packing_strategy()

        unplanned_item_list = SingleBinAddHeuristic(
            strategy,
            model.get_status_manager(),
            model.get_parameter()
        ).create_loading_plan(
            list(model.get_items()),
            container
        )

        # Put result into model
        model.set_containers([container])
        model.set_unplanned_items(unplanned_item_list)
